include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -673,241 +518,324 @@ void runNormalPrecompStream(in instr_stream_t stream, in mat2 dUV)
 {
-#ifndef NO_TWOSIDED
-	bool ts = false;
-#endif
+	setCurrInteraction(precomp);
 	for (uint i = 0u; i < stream.count; ++i)
 	{
 		instr_t instr = irr_glsl_MC_fetchInstr(stream.offset+i);
-#ifndef NO_TWOSIDED
-		handleTwosided(ts, instr);
+		uint op = instr_getOpcode(instr);
+
+		bool skip = (i == generator_offset);
+		// skip deltas
+#ifdef OP_THINDIELECTRIC
+		skip = skip || (op == OP_THINDIELECTRIC);
+#endif
+#ifdef OP_DELTATRANS
+		skip = skip || (op == OP_DELTATRANS);
+#endif
+
+		instr_eval_and_pdf_execute(instr, precomp, s, microfacet, skip);
+
+#if defined(OP_SET_GEOM_NORMAL)||defined(OP_BUMPMAP)
+		if (
+#ifdef OP_SET_GEOM_NORMAL
+			op == OP_SET_GEOM_NORMAL
+#ifdef OP_BUMPMAP
+			||
+#endif
+#endif
+#ifdef OP_BUMPMAP
+			op == OP_BUMPMAP
+#endif
+			) {
+			updateLightSampleAfterNormalChange(s);
+			updateMicrofacetCacheAfterNormalChange(s, microfacet);
+		}
 #endif
-		instr_remainder_and_pdf_execute(instr, s);
 	}
 
 	eval_and_pdf_t eval_and_pdf = readReg4(0u);
-	_out_pdf = eval_and_pdf.w;
-	return eval_and_pdf.rgb/eval_and_pdf.w;
+	return eval_and_pdf;
 }
 
-irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand)
+irr_glsl_LightSample irr_bsdf_cos_generate(in MC_precomputed_t precomp, in instr_stream_t stream, in vec3 rand, out vec3 out_remainder, out float out_pdf, out irr_glsl_AnisotropicMicrofacetCache out_microfacet, out uint out_gen_rnpOffset, out bool out_invalid_microfacet)
 {
 	uint ix = 0u;
 	instr_t instr = irr_glsl_MC_fetchInstr(stream.offset);
 	uint op = instr_getOpcode(instr);
+	vec3 u = rand;
+
+	out_pdf = 1.0;
+
+	setCurrInteraction(precomp);
+	bool is_coat = false;
 	while (!op_isBXDF(op))
 	{
 #ifdef OP_BLEND
 		if (op==OP_BLEND) {
 			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
 			params_t params = instr_getParameters(instr, bsdf_data);
-			float w = params_getBlendWeight(params);
+			vec3 w_ = params_getBlendWeight(params);
+			float w = colorToScalar(w_);
 			float rcpChoiceProb;
-			bool choseRight = irr_glsl_partitionRandVariable(w, rand.z, rcpChoiceProb);
+			bool choseRight = irr_glsl_partitionRandVariable(w, u.z, rcpChoiceProb);
 
 			uint right = instr_getRightJump(instr);
 			ix = choseRight ? right:(ix+1u);
+			out_pdf /= rcpChoiceProb;
 		}
 		else 
 #endif //OP_BLEND
+#ifdef OP_COATING
+		if (op==OP_COATING) {
+			bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
+			vec3 eta = bsdf_data_decodeIoR(bsdf_data, OP_COATING)[0];
+			vec3 fresnel = irr_glsl_fresnel_dielectric_frontface_only(eta, currInteraction.isotropic.NdotV);
+			float w = colorToScalar(fresnel);
+			float rcpChoiceProb;
+			bool choseCoated = irr_glsl_partitionRandVariable(w, u.z, rcpChoiceProb);
+
+			uint coated_ix = instr_getRightJump(instr);
+			ix = choseCoated ? coated_ix:(ix+1u);
+			out_pdf /= rcpChoiceProb;
```

Comment:
The COAT and BLEND only diverge in how they compute the RGB weight

after you get the weight, the rest of the operations are the same (RGB to monochrome, partitioning, pdf calc, jump calc)